### PR TITLE
fix(clustering) more reliable way to calculate data plane config hash

### PR DESCRIPTION
**Pull request by bungle**
### Summary

Previously data plane hash calculation was flaky and with recent OpenResty changes it become
even flakier. This PR tries to calculate data plane config hash in a more reliable way by
sorting the data and then calculating the hash out of it.

### CODE REVIEW

File: kong/clustering/data_plane.lua [hunk 1]
```diff
@@ -62,11 +65,35 @@ function _M.new(parent)
...
+      o[i+2] = ":"
+      o[i+3] = sort(v)
+      o[i+4] = ";"
+      i=i+4
+    end
+
+    return concat(o, nil, 1, i)
+  end
+
+  return tostring(t)
```
Comment (kikito) [inline kong/clustering/data_plane.lua:83]: Don't encode the number `1` the same as the string `"1"`. Also avoid problems with strings starting with `{` (conflict with table separators)
```suggestion
  if type(t) == "string" then
    return '"' .. t:gsub('"', '\\"') .. '"'
  else
    return tostring(t)
  end
```

**Review by kikito — CHANGES_REQUESTED** [review]

File: kong/clustering/data_plane.lua [hunk 2]
```diff
@@ -343,8 +378,6 @@ function _M:communicate(premature)
               -- count is guaranteed to not exceed 1
               config_semaphore:post()
             end
-
-            send_ping(c)
```
Comment (fffonion) [inline kong/clustering/data_plane.lua:347]: why is this removed?

File: kong/clustering/data_plane.lua [hunk 3]
```diff
@@ -62,11 +65,35 @@ function _M.new(parent)
 end
 
 
+local function sort(t)
```
Comment (fffonion) [inline kong/clustering/data_plane.lua:68]: maybe we can have a basic performance on this sort function.
for example if we have a json with 5M of JSON, how long does it take to sort it?

File: kong/clustering/data_plane.lua (hunk 2 above)
Comment (bungle) [reply kong/clustering/data_plane.lua:347]: Because it sends hash in wrong place. It sends old hash, and we want it to send new hash.

File: kong/clustering/data_plane.lua (hunk 3 above)
Comment (bungle) [reply kong/clustering/data_plane.lua:68]: I can do it, but I would expect it to be minimal. Like not even close to all that schema validation etc. stuff that also happens there.

File: kong/clustering/data_plane.lua (hunk 2 above)
Comment (fffonion) [reply kong/clustering/data_plane.lua:347]: ok, let's include this and also the ping_immediately change in commit message (or ideally in seperate commits, either works better for you)

Comment (bungle) [reply kong/clustering/data_plane.lua:347]: It is now on separate commit.

**Comment by dndx** [comment]
As discussed offline, let's postpone this to after 2.5 FF since it is a fix and can be in after FF.

File: kong/clustering/data_plane.lua (hunk 3 above)
Comment (bungle) [reply kong/clustering/data_plane.lua:56]: Here some results with 68MB JSON file:

```
(feat/lax-version-checks *) bungle@pro:~/Sources/kong$ ll config.json
-rw-r--r--  1 bungle  staff    68M Jun 23 10:45 config.json
(feat/lax-version-checks *) bungle@pro:~/Sources/kong$ resty perf.lua 
7e111baf0b58b2362b6646d29ca5c458
took: 2.3029999732971 secs
```

And here is the `perf.lua`:
```lua
local cjson = require "cjson"
local tablex = require "pl.tablex"
local concat = table.concat
local tostring = tostring
local ngx_null = ngx.null
local md5 = ngx.md5
local open = io.open


local function sort(t)
  if t == ngx_null then
    return "/null/"
  end

  local typ = type(t)
  if typ == "table" then
    local i = 1
    local o = { "{" }
    for k, v in tablex.sort(t) do
      o[i+1] = sort(k)
      o[i+2] = ":"
      o[i+3] = sort(v)
      o[i+4] = ";"
      i=i+4
    end
    if i == 1 then
      i = i + 1
    end
    o[i] = "}"

    return concat(o, nil, 1, i)

  elseif typ == "string" then
    return '$' .. t .. '$'

  elseif typ == "number" then
    return '#' .. tostring(t) .. '#'

  elseif typ == "boolean" then
    return '?' .. tostring(t) .. '?'

  else
    return '(' .. tostring(t) .. ')'
  end
end


local function hash(t)
  return md5(sort(t))
end


local function read_file(path)
    local file = open(path, "rb")
    local content = file:read "*a"
    file:close()
    return content
end


local content = read_file("config.json")
local json = cjson.decode(content)


ngx.update_time()
local started = ngx.now()
print(hash(json))
ngx.update_time()
local ended = ngx.now()
print("took: ", ended - started, " secs")
```

**Comment by bungle** [comment]
> As discussed offline, let's postpone this to after 2.5 FF since it is a fix and can be in after FF.

I don't think it needs postponed, just that it doesn't need to go to feature freeze. It has plenty of time to go to `2.5` final.

File: kong/clustering/data_plane.lua (hunk 3 above)
Comment (bungle) [reply kong/clustering/data_plane.lua:56]: With 5MB `json` it takes on my `2010` Mac: 0.2 secs.

Comment (fffonion) [reply kong/clustering/data_plane.lua]: Awesome!

File: kong/clustering/init.lua [hunk 4]
```diff
@@ -5,10 +5,54 @@ local pl_file = require("pl.file")
...
+local tostring = tostring
+local assert = assert
+local concat = table.concat
+local sort = table.sort
+local type = type
 
 
 local MT = { __index = _M, }
 
+local function table_to_sorted_string(t)
```
Comment (fffonion) [inline kong/clustering/init.lua:19]: let's have some basic test case for this function, just to catch any potential
behaviour change in LuaJIT or penlight for example.

**Review by fffonion — APPROVED** [review]
